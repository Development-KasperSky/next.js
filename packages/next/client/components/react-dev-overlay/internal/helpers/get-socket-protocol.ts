export function getSocketProtocol(assetPrefix: string): string {
  let protocol = window.location.protocol

  try {
    // assetPrefix is a url
    protocol = new URL(assetPrefix).protocol
  } catch (_) {}

  return protocol === 'http:' ? 'ws' : 'wss'
}
