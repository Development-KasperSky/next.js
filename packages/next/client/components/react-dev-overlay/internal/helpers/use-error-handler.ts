import { useEffect, useRef } from 'react'

export function useErrorHandler(
  handleOnUnhandledError: (event: WindowEventMap['error']) => void,
  handleOnUnhandledRejection: (
    event: WindowEventMap['unhandledrejection']
  ) => void
) {
  const stacktraceLimitRef = useRef<undefined | number>()

  useEffect(() => {
    try {
      const limit = Error.stackTraceLimit
      Error.stackTraceLimit = 50
      stacktraceLimitRef.current = limit
    } catch {}

    window.addEventListener('error', handleOnUnhandledError)
    window.addEventListener('unhandledrejection', handleOnUnhandledRejection)
    return () => {
      if (stacktraceLimitRef.current !== undefined) {
        try {
          Error.stackTraceLimit = stacktraceLimitRef.current
        } catch {}
        stacktraceLimitRef.current = undefined
      }

      window.removeEventListener('error', handleOnUnhandledError)
      window.removeEventListener(
        'unhandledrejection',
        handleOnUnhandledRejection
      )
    }
  }, [handleOnUnhandledError, handleOnUnhandledRejection])
}
