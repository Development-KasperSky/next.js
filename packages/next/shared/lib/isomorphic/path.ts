const path = process.browser
  ? require('next/dist/compiled/path-browserify')
  : require('path')

export default path
