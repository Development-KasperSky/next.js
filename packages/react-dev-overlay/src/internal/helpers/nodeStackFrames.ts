import { parse, StackFrame } from 'stacktrace-parser'

export function getFilesystemFrame(frame: StackFrame): StackFrame {
  const f: StackFrame = { ...frame }

  if (typeof f.file === 'string') {
    if (
      // Posix:
      f.file.startsWith('/') ||
      // Win32:
      /^[a-z]:\\/i.test(f.file) ||
      // Win32 UNC:
      f.file.startsWith('\\\\')
    ) {
      f.file = `file://${f.file}`
    }
  }

  return f
}

const symbolError = Symbol('NextjsError')

export function getErrorSource(error: Error): 'server' | 'edge-server' | null {
  return (error as any)[symbolError] || null
}

export function getServerError(
  error: Error,
  type: 'edge-server' | 'server'
): Error {
  let n: Error
  try {
    throw new Error(error.message)
  } catch (e) {
    n = e as Error
  }

  n.name = error.name
  try {
    n.stack = parse(error.stack!)
      .map(getFilesystemFrame)
      .map((f) => {
        let str = `    at ${f.methodName}`
        if (f.file) {
          let loc = f.file
          if (f.lineNumber) {
            loc += `:${f.lineNumber}`
            if (f.column) {
              loc += `:${f.column}`
            }
          }
          str += ` (${loc})`
        }
        return str
      })
      .join('\n')
  } catch {
    n.stack = error.stack
  }

  Object.defineProperty(n, symbolError, {
    writable: false,
    enumerable: false,
    configurable: false,
    value: type,
  })
  return n
}
