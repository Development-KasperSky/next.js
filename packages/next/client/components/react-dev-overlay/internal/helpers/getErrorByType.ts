import {
  ACTION_UNHANDLED_ERROR,
  ACTION_UNHANDLED_REJECTION,
} from '../error-overlay-reducer'
import { SupportedErrorEvent } from '../container/Errors'
import { getErrorSource } from './nodeStackFrames'
import { getOriginalStackFrames, OriginalStackFrame } from './stack-frame'

export type ReadyRuntimeError = {
  id: number
  runtime: true
  error: Error
  frames: OriginalStackFrame[]
}

export async function getErrorByType(
  ev: SupportedErrorEvent
): Promise<ReadyRuntimeError> {
  const { id, event } = ev
  switch (event.type) {
    case ACTION_UNHANDLED_ERROR:
    case ACTION_UNHANDLED_REJECTION: {
      return {
        id,
        runtime: true,
        error: event.reason,
        frames: await getOriginalStackFrames(
          event.frames,
          getErrorSource(event.reason),
          event.reason.toString()
        ),
      }
    }
    default: {
      break
    }
  }
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const _: never = event
  throw new Error('type system invariant violation')
}
