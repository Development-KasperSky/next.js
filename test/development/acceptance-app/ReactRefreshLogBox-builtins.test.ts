import { sandbox } from './helpers'
import { createNext, FileRef } from 'e2e-utils'
import { NextInstance } from 'test/lib/next-modes/base'
import path from 'path'

// TODO-APP: Investigate snapshot mismatch
describe('ReactRefreshLogBox app', () => {
  if (process.env.NEXT_TEST_REACT_VERSION === '^17') {
    it('should skip for react v17', () => {})
    return
  }

  let next: NextInstance

  beforeAll(async () => {
    next = await createNext({
      files: new FileRef(path.join(__dirname, 'fixtures', 'default-template')),
      dependencies: {
        react: '0.0.0-experimental-9cdf8a99e-20221018',
        'react-dom': '0.0.0-experimental-9cdf8a99e-20221018',
      },
      skipStart: true,
    })
  })
  afterAll(() => next.destroy())

  // Module trace is only available with webpack 5
  test('Node.js builtins', async () => {
    const { session, cleanup } = await sandbox(
      next,
      new Map([
        [
          'node_modules/my-package/index.js',
          `
          const dns = require('dns')
          module.exports = dns
        `,
        ],
        [
          'node_modules/my-package/package.json',
          `
          {
            "name": "my-package",
            "version": "0.0.1"
          }
        `,
        ],
      ])
    )

    await session.patch(
      'index.js',
      `
      import pkg from 'my-package'

      export default function Hello() {
        return (pkg ? <h1>Package loaded</h1> : <h1>Package did not load</h1>)
      }
    `
    )
    expect(await session.hasRedbox(true)).toBe(true)
    expect(await session.getRedboxSource()).toMatchSnapshot()

    await cleanup()
  })

  test('Module not found', async () => {
    const { session, cleanup } = await sandbox(next)

    await session.patch(
      'index.js',
      `import Comp from 'b'
      export default function Oops() {
        return (
          <div>
            <Comp>lol</Comp>
          </div>
        )
      }
    `
    )

    expect(await session.hasRedbox(true)).toBe(true)

    const source = await session.getRedboxSource()
    expect(source).toMatchSnapshot()

    await cleanup()
  })

  test('Module not found empty import trace', async () => {
    const { session, cleanup } = await sandbox(next)

    await session.patch(
      'app/page.js',
      `'use client'
      import Comp from 'b'
      export default function Oops() {
        return (
          <div>
            <Comp>lol</Comp>
          </div>
        )
      }
    `
    )

    expect(await session.hasRedbox(true)).toBe(true)

    const source = await session.getRedboxSource()
    expect(source).toMatchSnapshot()

    await cleanup()
  })

  test('Module not found missing global CSS', async () => {
    const { session, cleanup } = await sandbox(
      next,
      new Map([
        [
          'app/page.js',
          `'use client'
          import './non-existent.css'
        export default function Page(props) {
          return <p>index page</p>
        }
      `,
        ],
      ])
    )
    expect(await session.hasRedbox(true)).toBe(true)

    const source = await session.getRedboxSource()
    expect(source).toMatchSnapshot()

    await session.patch(
      'app/page.js',
      `'use client'
      export default function Page(props) {
        return <p>index page</p>
      }
    `
    )
    expect(await session.hasRedbox(false)).toBe(false)
    expect(
      await session.evaluate(() => document.documentElement.innerHTML)
    ).toContain('index page')

    await cleanup()
  })
})
