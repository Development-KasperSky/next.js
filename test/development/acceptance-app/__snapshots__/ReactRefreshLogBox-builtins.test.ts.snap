// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactRefreshLogBox app Module not found 1`] = `
"./index.js:1:0
Module not found: Can't resolve 'b'
> 1 | import Comp from 'b'
  2 |       export default function Oops() {
  3 |         return (
  4 |           <div>

Import trace for requested module:
./app/page.js

https://nextjs.org/docs/messages/module-not-found"
`;

exports[`ReactRefreshLogBox app Module not found empty import trace 1`] = `
"./app/page.js:2:6
Module not found: Can't resolve 'b'
  1 | 'use client'
> 2 |       import Comp from 'b'
    |      ^
  3 |       export default function Oops() {
  4 |         return (
  5 |           <div>

https://nextjs.org/docs/messages/module-not-found"
`;

exports[`ReactRefreshLogBox app Module not found missing global CSS 1`] = `
"./app/page.js:2:10
Module not found: Can't resolve './non-existent.css'
  1 | 'use client'
> 2 |           import './non-existent.css'
    |          ^
  3 |         export default function Page(props) {
  4 |           return <p>index page</p>
  5 |         }

https://nextjs.org/docs/messages/module-not-found"
`;

exports[`ReactRefreshLogBox app Node.js builtins 1`] = `
"./node_modules/my-package/index.js:2:0
Module not found: Can't resolve 'dns'

Import trace for requested module:
./index.js
./app/page.js

https://nextjs.org/docs/messages/module-not-found"
`;
